mutation_writer/token_group_based_splitting_writer.cc

```diff
@@ -45,8 +45,8 @@ class token_group_based_splitting_mutation_writer {
 
         if (_current_writer && _current_group_id > prev_group_id) [[unlikely]] {
             _current_writer->consume_end_of_stream();
-            return _current_writer->close().then([this] {
-                _current_writer = std::nullopt;
+            bucket_writer_v2 wr = std::move(*std::exchange(_current_writer, std::nullopt));
+            return wr.close().then([this, wr = std::move(wr)] {
```

Comment:
Is this safe in general? If close was implemented as do_something_then([this] { update_some_local_state }), then it would be accessing a freed region. Maybe we can fix it while switching to coroutine. Coroutine might introduce one allocation though, and for small-sized partitions might have some cost. Since this is in slow path, switching to next writer, maybe we could wrap wr in a unique ptr to guarantee stable address.

Reply:
Penalizing the common path with a coroutine doesn't make sense.  We can have a coroutine for closing the current writer or use do_with to keep the object alive until close is resolved. But bucket_writer_v2 doesn't really need that as far as I could see.
I'll double check that (or you can too, cause it's getting late here)

Reply:
Yes, I thought first of coroutine, but later dropped the idea while writing as you can see. I think it's dangerous to call a method of an object and destroy it in the middle of execution. One change the implementation of close in the future (I didn't really look into it but it doesn't matter since need to address the fragility), and we have a bug. Either wrap wr in a unique ptr when closing writer (it's slow path since there will be much more partitions than split boundaries), or change _current_writer to be unique_ptr instead of optional.

Reply:
Yeah, keeping the writer in a unique_ptr would work well.

Reply:
It's better to do it in a coroutine, for readability. maybe_switch_to_new_writer() can exit quickly if it doesn't need to do anything, and delegate to a coroutine if it does.

Reply:
> It's better to do it in a coroutine, for readability. maybe_switch_to_new_writer() can exit quickly if it doesn't need to do anything, and delegate to a coroutine if it does.

Done in v2 ([3936166](https://github.com/scylladb/scylladb/commit/39361668416edee2f4a16252113d664e862f070b))